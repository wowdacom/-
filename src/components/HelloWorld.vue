<script setup>
import { ref, onMounted } from "vue";
import SINGERS_INVOICES from "../assets/singerInvoice.json";
import SINGERS from "../assets/singerList.json";

defineProps({
  msg: String,
});

const alert = ref("");

const singer_statement = (invoice, singer) => {
  let msg = "";
  let originalTotalAmount = 0;
  let totalAmount = 0;
  let totalDiscount = 0;

  for (let ticket of invoice.performances) {
    let cost = 0;
    let save = 0;
    let off = 1;
    switch (singer[ticket.playID].type) {
      case "concert":
        off = 0.9;
        cost = ticket.fee * off;
        save = ticket.fee - cost;
        break;
      case "exhibition":
        off = 0.5;
        cost = ticket.fee * off;
        save = ticket.fee - cost;
        break;
      default:
        throw new Error(`unkown type: ${singer[ticket.playID].type}`);
    }

    originalTotalAmount += ticket.fee;
    totalAmount += cost;
    totalDiscount += save;
  }

  msg += `<p>${
    invoice.customer
  } 您好：\n本次總票價為 <span style="text-decoration:line-through">${originalTotalAmount.toLocaleString(
    "en-US"
  )}$</span></p><br>`;
  msg += `<p>實際消費為 ${totalAmount.toLocaleString("en-US")}$</p>`;
  msg += `<p>此次會員資格為您節省了 ${totalDiscount.toLocaleString(
    "en-US"
  )}$</p>`;
  return msg;
};

onMounted(() => {
  const message = singer_statement(SINGERS_INVOICES[0], SINGERS);
  // Try edit me
  console.log(message);

  // Update header text
  alert.value = message;
});
</script>

<template>
  <p class="text-red-500 text-center">
    <span class="text-red-200">✦✦✦✦ </span>2022
    年底前申辦會員，享有演唱會九折優惠，一般展覽一律半價，敬啟把握!<span></span>
    <span class="text-red-200"> ✦✦✦✦</span>
  </p>
  <br />
  <h1 v-html="alert" id="header"></h1>
</template>

<style scoped>
h1 {
  background: transparent;
  color: #fcbe24;
  padding: 0 24px;
  display: flex;
  justify-content: center;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen,
    Ubuntu, Cantarell, "Open Sans", "Helvetica Neue", sans-serif;
  flex-direction: column;
  white-space: pre-wrap;
}
</style>
