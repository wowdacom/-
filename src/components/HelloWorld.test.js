import { describe, it, expect, test } from "vitest";
import { nextTick } from "vue";
import { mount } from "@vue/test-utils";
import HelloWorld from "./HelloWorld.vue";

//確認可以執行測試
describe("Basic Test", () => {
  it("1+1=2", () => {
    expect(1 + 1).toBe(2);
  });
});

test("total", async () => {
  const wrapper = mount(HelloWorld, {
    props: {
      customer: "廣東粥",
    },
  });
  await nextTick();

  expect(wrapper.find('[data-test="target"]').html()).toContain("5,000");
});

test("discount", async () => {
  const wrapper = mount(HelloWorld, {
    props: {
      customer: "廣東粥",
    },
  });
  await nextTick();

  expect(wrapper.find('[data-test="target"]').html()).toContain("500");
});

// describe("Hellow World", () => {
//   it("Hellow World", async () => {
//     const wrapper = mount(HelloWorld, {
//       props: {
//         customer: "東區桂綸鎂",
//       },
//     });
//     await nextTick();
//     expect(wrapper.text()).toContain("東區桂綸鎂");
//   });
// });

//注意 render 的順序

//參考資料：
// ShawnL https://ithelp.ithome.com.tw/articles/10303784
// Mike https://ithelp.ithome.com.tw/articles/10282236?sc=iThomeR

//設計我要做的測試，確定修改後不會壞掉

// describe("確認算出來的數字正確", () => {
//   it("Hellow World", async () => {
//     const wrapper = mount(HelloWorld, {
//       props: {
//         customer: "公館陳柏霖",
//       },
//     });
//     await nextTick();
//     expect(wrapper.text()).toContain("395");
//   });
// });

// test("snapshot", async () => {
//   const wrapper = mount(HelloWorld, {
//     props: {
//       customer: "公館陳柏霖",
//     },
//   });
//   await nextTick();
//   expect(wrapper.find('[data-test="target"]').html()).toMatchSnapshot();
// });

//分解 Statement 函式

//移除 play 變數

// 取出 volume credit

// 移除 format 變數

// 移除總 volume credit

// 狀態：有許多內嵌的函示

//將計算階段與格式化階段拆開

// 狀態：拆成兩個檔案(與階段)

//用類型辨識計算 /建立 Performance 計算器 /將函式移入 calculator /讓 performance calculator 是多型的

// 狀態：用多型的 claculator 建立資料

// 小結論
