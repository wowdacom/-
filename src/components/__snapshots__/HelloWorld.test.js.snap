// Vitest Snapshot v1

exports[`snapshot 1`] = `
"<h1 data-test=\\"target\\" id=\\"header\\" data-v-e17ea971=\\"\\">
  <p>公館陳柏霖 親愛的消費者您好：
    本次總票價為 <span style=\\"text-decoration:line-through\\">2,550$</span></p><br>
  <p>實際消費為 2,155$</p>
  <p>此次會員資格為您節省了 395$</p>
</h1>"
`;
